salvo/ENVOY_DEVELOP_WORKFLOW.md

```diff
@@ -0,0 +1,135 @@
+
+    Then Salvo will build each Envoy commit and the baseline version. You can add bazel options in `source.bazelOptions`, and they will be added as parameters to bazel command when building Envoy.  
+
+    Finally you can also determine which Nighthawk version to use like Envoy's configuration, in this section we will build the latest Nighthawk version. You can find more details about configuration fields from [APIs definition](./api). 
+
+6. Details of test cases
+
+    The Benchmark Framework is developed via PyTest Framework. Therefore, the test cases file is defined as a python file contains the test cases. The [default test cases file](https://github.com/envoyproxy/nighthawk/blob/main/benchmarks/test/test_discovery.py) provides an example to configure Nighthawk client/server, outputs and test cases.
+
+    There are two main parts in the test case file.
+
+    - `_run_benchmark` function: 
+
+      At first, it defined a function named [`_run_benchmark`](https://github.com/envoyproxy/nighthawk/blob/main/benchmarks/test/test_discovery.py#L20) to run the specific PyTest fixture, which will define the behavior of Envoy and [Nighthawk test server](https://github.com/envoyproxy/nighthawk/blob/main/source/server/README.md) to be tested, you can find fixture definitions from these two files:
+
+      https://github.com/envoyproxy/nighthawk/blob/main/benchmarks/envoy_proxy.py
+
+      https://github.com/envoyproxy/nighthawk/blob/main/test/integration/integration_test_fixtures.py
+    
+      Then, the `_run_benchmark` function provides some configurable parameters used by Nighthawk client to send test requests, such as `rps`, `duration`, `request_body_size` and so on. By default, Nighthawk will use closed-loop mode, because closed-loop testing can be useful to get a sense of the maximum requests per second that can be squeezed out of a system (irrespective of the high latency incurred), as well as for comparing numbers with other load generators that use this methodology. More details about closed-loop mode: https://github.com/envoyproxy/nighthawk/blob/main/docs/root/terminology.md#closed-loop
+
+      Finally, `_run_benchmark` function defines the format of result outputs after the benchmark is done.
+
+    -  Test cases functions: 
+
+        After `_run_benchmark` function defined, you can write the specific test cases to use it. Take [`test_http_h1_small_request_small_reply_via`](https://github.com/envoyproxy/nighthawk/blob/8d6947b2d4b41edf0ea95a811d129c7a81629b86/benchmarks/test/test_discovery.py#L87) as an example, it will setup a Nighthawk client as the frontend, the Envoy Proxy to be tested as the load balancer, and a Nighthawk test server as the backend. The configurations of Envoy and test server can be assigned with Python decorators. Then you can invoke `_run_benchmark` function with specific fixture and parameters. 
```

Comment:
I see that we are talking about the setup and the roles of the three components (Nighthawk, Envoy, Test Server). In relation to my previous comment - this seems a bit too late and it isn't technically correct. Since the test function which we are describing isn't the one doing the setup, the test fixture that we mentioned before is.

Can we extract the portion that describes the test architecture and mention it early on before we dive into the details (as noted in the other comment).

Reply:
I agree with you that it's too late to introduce the three key components here, can I add the introduction and architecture diagram in README.md and link it here?

Reply:
I think that is a good idea. I would create a separate document called `test_architecture.md` and link to it from `README.md` to keep the `README.md` relatively small and focused.

Reply:
Yes, the separate document called `test_architecture.md` is better. I can link to it after your commit.

Reply:
Sorry my comment was meant as a suggestion for you. I am currently not planning to create that document and was hoping you could.

I won't be able to directly contribute to Salvo beyond code reviews until I finalize my current projects.

Reply:
I can understand, I'll create it.

Reply:
Thank you @gyohuangxin !